src/openslide-vendor-dicom.c

```
@@ -0,0 +1,840 @@
+    bool found_difference = false;
+
+    for (int j = 0; j < 4; j++) {
+      const char *type;
+      if (!get_tag_str(f->metadata, "ImageType", j, &type)) {
+        return false;
+      }
+
+      if (!g_str_equal(type, types->types[i][j])) {
+        found_difference = true;
+        break;
+      }
+    }
+
+    if (!found_difference) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+static bool associated_get_argb_data(struct _openslide_associated_image *_img,
+                                     uint32_t *dest,
+                                     GError **err) {
+  struct associated *a = (struct associated *) _img;
+
+  DcmError *dcm_error = NULL;
+  DcmFrame *frame = dcm_filehandle_read_frame(&dcm_error,
+                                              a->file->filehandle,
+                                              a->file->metadata,
+                                              a->file->bot,
+                                              1);
+  if (frame == NULL) {
+    dicom_propagate_error(err, dcm_error);
+    return false;
+  }
+
+  print_frame(frame);
+
+  return _openslide_jpeg_decode_buffer(dcm_frame_get_value(frame),
+                                       dcm_frame_get_length(frame),
+                                       dest,
+                                       a->base.w, a->base.h,
+                                       err);
+}
+
+static void associated_destroy(struct associated *a) {
+  dicom_file_destroy(a->file);
+  g_slice_free(struct associated, a);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(associated, associated_destroy)
+
+static void openslide_associated_destroy(struct _openslide_associated_image *_img) {
+  struct associated *a = (struct associated *) _img;
+  associated_destroy(a);
+}
+
+static const struct _openslide_associated_image_ops dicom_associated_ops = {
+  .get_argb_data = associated_get_argb_data,
+  .destroy = openslide_associated_destroy
+};
+
+static bool read_whole_file(struct dicom_file *f) {
```

We should probably propagate errors to the caller, which can decide what to do with them (such as logging under a debug flag).  In general, functions should propagate any errors, unless they're a convenience function e.g. `get_tag_int` that has a reason not to.

---

We have to call `read_whole_file()` before we can get any of the main metadata, which we have to do decide if this file is one of the things we can handle. So I think this is more of a tester (like `get_tag`) than an action the loader is performing that can fail.